spec/rubocop/cop/style/or_assignment_spec.rb

```diff
@@ -0,0 +1,321 @@
+# frozen_string_literal: true
+
+describe RuboCop::Cop::Style::OrAssignment do
+  let(:config) { RuboCop::Config.new }
+  subject(:cop) { described_class.new(config) }
+
+  context 'when using var = var ? var : something' do
+    it 'registers an offense with normal variables' do
+      expect_offense(<<-RUBY.strip_indent)
+        foo = foo ? foo : 'default'
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ Use the double pipe equals operator `||=` instead.
+      RUBY
+    end
+
+    it 'registers an offense with instance variables' do
+      expect_offense(<<-RUBY.strip_indent)
+        @foo = @foo ? @foo : 'default'
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Use the double pipe equals operator `||=` instead.
+      RUBY
+    end
+
+    it 'registers an offense with class variables' do
+      expect_offense(<<-RUBY.strip_indent)
+        @@foo = @@foo ? @@foo : 'default'
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Use the double pipe equals operator `||=` instead.
+      RUBY
+    end
+
+    it 'registers an offense with global variables' do
+      expect_offense(<<-RUBY.strip_indent)
+        $foo = $foo ? $foo : 'default'
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Use the double pipe equals operator `||=` instead.
+      RUBY
+    end
+
+    it 'autocorrects normal variables to `var ||= something`' do
+      expect(autocorrect_source('x = x ? x : 3')).to eq('x ||= 3')
+    end
+
+    it 'autocorrects instance variables to `var ||= something`' do
+      expect(autocorrect_source('@x = @x ? @x : 3')).to eq('@x ||= 3')
+    end
+
+    it 'autocorrects class variables to `var ||= something`' do
+      expect(autocorrect_source('@@x = @@x ? @@x : 3')).to eq('@@x ||= 3')
+    end
+
+    it 'autocorrects global variables to `var ||= something`' do
+      expect(autocorrect_source('$x = $x ? $x : 3')).to eq('$x ||= 3')
+    end
+
+    it 'does not register an offense if any of the variables are different' do
+      expect_no_offenses('foo = bar ? foo : 3')
+      expect_no_offenses('foo = foo ? bar : 3')
+    end
+  end
+
+  context 'when using var = if var; var; else; something; end' do
+    it 'registers an offense with normal variables' do
+      inspect_source(<<-RUBY.strip_indent)
+        foo = if foo
+                foo
+              else
+                'default'
+              end
+      RUBY
+      expect(cop.offenses.size).to eq(1)
```

Comment:
Checking for offense count is a little redundant because of the check for the message. The number of offenses will match the number off messages.

Testing `cop.highlights` would be beneficial.